### ISSUE TITLE
MediaStreamTracks are ended after returning from sleep

### ISSUE THREAD

**Issue by a-type** [twilio/twilio-video.js#1447]
- [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**

- Run the latest video quickstart example
- If relevant, I'm on MacOS 11.2.3 using a Macbook Pro 13-inch with Core i5 (2020)
- Join with 2 participants, both sharing audio/video
- Put the computer to sleep for any length of time
- Restore the computer from sleep

**Expected behavior:**

- When Twilio reconnects after device sleep, either:
  - Track publications are restored and reconnected
  - OR, track publications fail and are cancelled or otherwise emit errors

**Actual behavior:**

The room reconnects after sleep and track publications are re-established, but the underlying `MediaStreamTrack` objects have `readyState: 'ended'` and no media is transmitted to remote participants. Each participant sees their own media stream as active, but peers' media streams as empty (black video).

`room.state` is `connected`.

[Here](https://gist.github.com/a-type/7c43dffd4c7c38ebb1fdf12e9cdb8b85) is a full log from one of the peers. Included are some expression evaluations which indicate that the remote video publication has `isSubscribed: true, isTrackEnabled: true`, the track has `isEnabled: true`, and the MediaStreamTrack has `readyState: ended`. However, when inspecting the corresponding local stream track (which the other peer client sees as `readyState: ended`), the `readyState` is `live` (as seen on the last line).

No error events are emitted. From a client perspective it appears as if the room reconnection was successful and all local tracks are published and streaming, but all peers receive empty tracks and the client cannot see or hear anyone else despite receiving their publications.

**Software versions:**

 - [x] Browser(s): Chrome 89.0.4389.114
 - [x] Operating System: MacOS 11.2.3
 - [x] twilio-video.js: 2.13.1
 - [x] Third-party libraries (e.g., Angular, React, etc.):

If anyone can find a workaround or way to detect this bad state I'd be fine with that. The browser does not give any indication of power state changes that I'm aware of, and I can't use the page visibility API as I don't want to assume a failure if the user just switches tabs or minimizes the app.

I've tested reconnections in other scenarios, including loss of internet (via disabling the device or poor connection), and they don't reproduce this behavior. It seems consistently tied to cycling sleep. While this may seem like an edge case, our app has long-lived sessions and we regularly see users sleep their computers while a tab is open. In these cases users may return to the tab later and be confused when A/V is (silently) not working or may not even notice the problem until later.

**Comment by a-type** [comment]
Because this issue is a little hard to describe well I've recorded videos of it.

https://youtu.be/qKrDZfiS1xo

The video shows the full behavior, starting with connecting two clients in the demo app, then putting the computer to sleep and waking it. I then walk through the two behaviors I see: one for clients which come back from sleep, and one when new clients connect to them and try to share tracks.

**Comment by makarandp0** [comment]
Hello @a-type, Thank you for the detailed description and the video. I was able to repro this issue. On chrome I notice that track goes to the "ended" state very quickly after sleep.  It  stays in "live" state in both firefox and safari if the computer was quickly woken from the sleep.

I have [filed a bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=12655) for chrome. 

One workaround that comes to mind is to watch for RemoteVideoTrack going into "ended" state and then communicate to publisher using data channels that the track was "ended", and they need to unpublish and republish.

Thanks,
Makarand

**Comment by a-type** [comment]
Good idea, thanks for narrowing it down. I'll see about coordinating that.

**Comment by a-type** [comment]
Oh, much easier! Thanks, I'll probably use that workaround and incorporate it into our extensions of the automated reconnection logic.
